### ISSUE TITLE
Handling inverted photon order

### ISSUE THREAD

**Issue by smXplorer** [Photon-HDF5/phconvert#31]
Some electronics (e.g. our SPC-630 TCSPC board) sometimes transfer photon data in the wrong order, resulting in macro timestamps appearing like so:

t_i, t_i+1,...,**t_i+p+1**,**t_i+p**,t_i+p+2,...

When decoding this series, what happens is that generally, t_i+p which now comes AFTER t_i+p+1, is smaller than t_i+p+1, and the phconvert code interprets this as a complete 2ˆ24x50 ns gap between the two photons.

This is wrong, but fortunately this "electronic feature" can be easily detected by testing whether two consecutive timestamps are separated by more than rollover/2, where rollover is the maximum value of the counter (in the case of SPC-630, 2ˆ24).

**Comment by tritemio** [comment]
If the board outputs out of order timestamps it is a problem. Can you past a small file showing the issue? (You can drag and drop it in the comment box here to make an attachment).

Looking at the phconvert's [code for decoding SPC files](https://github.com/Photon-HDF5/phconvert/blob/0.8.2/phconvert/bhreader.py#L98-L123), it seems that the overflow correction is based on the overflow bit, not on negative jumps detection. So you should end up with a file with a timestamp (`t_i+p` in your example) with one overflow too much. Basically you will have a negative jump just after a big positive jump. Is this what happens?

**Comment by smXplorer** [comment]
Yes. I forgot to warn you of this problem with that specific board (in addition to the fact that there are some bogus counts in unconnected channels every now and then). A file provided by Yung which shows this problem multiple times throughout the trace was shared with you on Google docs.
Not sure why that occurs, but this is unfortunately an issue with the board.
Thanks.

**Comment by tritemio** [comment]
I looked at the file and the problem is not timestamps out of order. I noticed that there are a lot of repeated timestamps, this is already weird since the board has a single TAC as far as I remember. Anyway, in some pair of identical timestamps the second has the overflow bit set, which is even more weird since the timestamp is the same. The result is overflow correction for the second timestamp and therefore a gap of exactly 2^24 units. Timestamps out of order would generate negative jumps but there are none.

See the attached notebook for the analysis of the issue and a quick fix:

[Overflow bug in SPC-630 board.ipynb.zip](https://github.com/Photon-HDF5/phconvert/files/2582734/Overflow.bug.in.SPC-630.board.ipynb.zip)

**Comment by smXplorer** [comment]
Yung mentioned that upon looking at the pairs of timestamps showing the rollover jump, they ended up having the same time stamp after rollover correction (which is what you are saying). I was not aware of this "feature" of the SPC-630 and like you, have no idea what it is due to. I believe the "backward" progression feature can be observed in other setups (maybe those involving the PCI-6602?). In any case, it is useful to make these checks when converting macrotimes, no matter what the hardware is, since these "impossible" cases are in fact observed in real life.

**Comment by tritemio** [comment]
This issue is weird. There are unanswered questions that point to some type of data corruptions:

1. why there are duplicated timestamps (in large numbers)?
2. why in some pairs of identical timestamps the second also have the overflow bit set?
3. are these real photons? What should we remove? Leave all timestamps even if these are unphysical counts? Remove all duplicate timestamps (each timestamp becomes unique)? Remove all timestamps that are not unique?

Without answers to these questions we cannot find a correct solution.

To conclude, I gave an ad-hoc solution in the previous notebook, also linked below for easy viewing:

https://gist.github.com/tritemio/ada9b8d413329cb5edbee942591f7e4d

In this solution, I don't remove timestamps, I leave them all (duplicated or not). I only correct the additional overflow. This could be easily integrated into phconvert (PR welcome). But this solution can be potentially "wrong" depending on the answers to the previous questions. Therefore, I don't think that we should always silently attempt to correct these gaps. Better leave this as an ad-hoc fix for troublesome boards.

**Comment by smXplorer** [comment]
I agree that removing timestamps is not the way to go. Leaving duplicates (without adding a rollover gap) is the fail-safe mechanism. It is easy to spot in post-processing, and as long as there is an explanation in phconvert of how this situation is dealt with, I don't think there is any loss of information. There is also no ambiguity (that it is a board problem), as there is no current detector that I know of that can return two simultaneous pulses and  those be processed by the same board.

Have you ever noticed a case where the second identical timestamp an overflow bit set to 1?

**Comment by tritemio** [comment]
@smXplorer, the issue is exactly that: the overflow bit is 1 in the second timestamp in a pair of identical ones. In other words, there are a lot of timestamps duplications (in this file ~3% of photons). When a duplication happens for an overflow photon the overflow correction is applied twice leading to gaps.

All these duplicated timestamps should be removed are they will certainly bias TCSPC decays and burst histograms.

I slightly edited the notebook above to make the issue clearer.

**Comment by smXplorer** [comment]
3% is problematic indeed, unless the nanotime is completely bogus (NaN, 0 or some other arbitrary value). I'll check the file again. What's the value of the "INVALID" bit in those cases?

**Comment by smXplorer** [comment]
Actually, I answered my own question: there is no invalid flag that I can see. However, there are GAP flags in some of the incriminated photons: those corresponds to a possible recording gap due to FIFO Full.
Checking the file, I see that the count rate reaches over 1 MHz, which can result in such a situation...
